include/pybind11/eigen.h

```diff
@@ -641,6 +643,343 @@ struct type_caster<Type, enable_if_t<is_eigen_other<Type>::value>> {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast_impl(&src, policy, parent);
+    }
+
+    static handle
+    cast(const Eigen::TensorMap<Type> &src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast(&src, policy, parent);
+    }
+
+    static handle cast(Eigen::TensorMap<Type> *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    static handle
+    cast(const Eigen::TensorMap<Type> *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    template <typename C>
+    static handle cast_impl(C *src, return_value_policy policy, handle parent) {
+        bool dec_parent = false;
+        bool writeable = false;
+        switch (policy) {
+            case return_value_policy::reference:
+                parent = none().release();
+                dec_parent = true;
+                writeable = !std::is_const<C>::value;
+                break;
+
+            case return_value_policy::reference_internal:
+                // Default should do the right thing
+                dec_parent = false;
+                writeable = !std::is_const<C>::value;
+                break;
+
+            default:
+                // move, take_ownership don't make any sense for a ref/map:
+                pybind11_fail("Invalid return_value_policy for Eigen Map type, must be either "
+                              "reference or reference_internal");
+        }
+
+        handle result = array_t<typename H::Element, eigen_to_numpy<H::Options>::flag>(
+                            H::get_shape(*src), src->data(), parent)
+                            .release();
+
+        if (dec_parent) {
+            parent.dec_ref();
+        }
+
+        if (!writeable) {
+            array_proxy(result.ptr())->flags &= ~detail::npy_api::NPY_ARRAY_WRITEABLE_;
+        }
+
+        return result;
+    }
+
+protected:
+    // TODO: Move to std::optional once std::optional has more support
+    std::unique_ptr<Eigen::TensorMap<Type>> value;
+
+public:
+    static constexpr auto name = const_name("Eigen::TensorMap");
```

Comment:
This is inconsistent, all the other casters in this file output the proper `numpy.ndarray[TYPE]` typing and this caster should too. This string determine what typing string the caster generates in Python is. Therefore, it should be be some form of numpy.ndarray etc.

Reply:
Done